server/src/main/java/org/elasticsearch/cluster/routing/allocation/NodeAllocationStatsCalculator.java

```diff
@@ -86,21 +96,20 @@ public Map<String, NodeAllocationAndClusterBalanceStats> stats(
                 forecastedWriteLoad += writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
                 forecastedDiskUsage += Math.max(indexMetadata.getForecastedShardSizeInBytes().orElse(0), shardSize);
                 currentDiskUsage += shardSize;
-
             }
-            float currentNodeWeight = weightFunction.nodeWeight(
+            float currentNodeWeight = weightFunction.calculateNodeWeight(
                 shards,
                 avgShardsPerNode,
                 forecastedWriteLoad,
                 avgWriteLoadPerNode,
                 currentDiskUsage,
                 avgDiskUsageInBytesPerNode
             );
-            stats.put(
+            nodeAllocationStats.put(
                 node.nodeId(),
-                new NodeAllocationAndClusterBalanceStats(
+                new NodeShardAllocationStats(
                     shards,
-                    desiredBalance != null ? undesiredShards : -1,
```

Comment:
There's no obvious reason why this should be 0 vs -1. We don't check for -1 value anywhere that I can see. If the desiredBalance is null, then undesiredShards is set to 0.

Reply:
It seems like it could be a useful distinction between "unknown" and "none"? but if we don't use that distinction anywhere I have no strong objections to removing it. @pxsalehi might have more context.

I wonder if the intention was to omit `-1`s when calculating `es.allocator.allocations.node.undesired_shard_count.current`

Reply:
If we keep the -1 as a magic number it should be documented in the javadoc.

Reply:
The reason is documented in our [public docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html). :) This is [existing code](https://github.com/elastic/elasticsearch/commit/e9f899ee6913fe00dc8ef7a4254c76e8dca31b47#diff-0eb22f36dd305443d2f3fbe16af5ae194ad9aa488072ad91f9a65e5c646c9591R103) and wasn't added recently. Just moved around. If there is no good reason for removing it, I think we should leave it in. Strictly speaking, we're changing (breaking?) a bit the meaning of a filed in the stats API. Maybe adding comment is more helpful, and more aligned with the PR scope.